# Personalised Cancer Diagnosis Results

# 5.1. Bag of Words Tabular Result

In [4]:
import pandas as pd
dataframe = pd.DataFrame(
{
    "Maching Learning Model": ['Naive Bayes','K' 'Nearest Neighbours','Logistic Regression [class balancing]','Logistic Regression [without class balancing]', 'Support Vector Machines', 'Random Forest [One Hot]', 'Random Forest [Response Coding]', 'Stacking [LR, SVM, NB] with LR', 'Stacking [LR,Linear,RF] with LR', 'Stacking [NB, LR, RF] with LR', 'Stacking [K-NN, NB, Linear] with LR', 'Stacking [LR, K-NN, Linear] with LR', 'Stacking [LR, Linear] with LR', 'Maximum Voting Classifier'],
    "Train (%)": [0.9,0.7,0.61,0.62,0.76,0.70,0.06,0.67,0.65,0.66,0.74,0.66,0.68,0.94],
    "CV (%)": [1.35,1.10,1.21,1.25,1.24,1.17,1.41,1.22,1.14,1.14,1.15,1.14,1.16,1.28],
    "Test (%)": [1.27,1.09,1.10,1.13,1.15,1.16,1.38,1.15,1.08,1.09,1.11,1.09,1.11,1.22],
    "Misclassification (%)": [44.36,39.66,38.34,37.21,39.47,38.53,51.87,37.29,35.33,35.33,34.58,36.09,36.39,38.19],
}

)
dataframe

Maching Learning Model  Train (%)  CV (%)  \
0                                     Naive Bayes       0.90    1.35   
1                             KNearest Neighbours       0.70    1.10   
2           Logistic Regression [class balancing]       0.61    1.21   
3   Logistic Regression [without class balancing]       0.62    1.25   
4                         Support Vector Machines       0.76    1.24   
5                         Random Forest [One Hot]       0.70    1.17   
6                 Random Forest [Response Coding]       0.06    1.41   
7                  Stacking [LR, SVM, NB] with LR       0.67    1.22   
8                 Stacking [LR,Linear,RF] with LR       0.65    1.14   
9                   Stacking [NB, LR, RF] with LR       0.66    1.14   
10            Stacking [K-NN, NB, Linear] with LR       0.74    1.15   
11            Stacking [LR, K-NN, Linear] with LR       0.66    1.14   
12                  Stacking [LR, Linear] with LR       0.68    1.16   
13                      Maximum Voting Classifier       0.94    1.28   

    Test (%)  Misclassification (%)  
0       1.27                  44.36  
1       1.09                  39.66  
2       1.10                  38.34  
3       1.13                  37.21  
4       1.15                  39.47  
5       1.16                  38.53  
6       1.38                  51.87  
7       1.15                  37.29  
8       1.08                  35.33  
9       1.09                  35.33  
10      1.11                  34.58  
11      1.09                  36.09  
12      1.11                  36.39  
13      1.22                  38.19

# 5.2 Term Frequency - Inverse Word Frequency Tabular Result

In [5]:
import pandas as pd
dataframe = pd.DataFrame(
{
    "Maching Learning Model": ['Naive Bayes','K' 'Nearest Neighbours','Logistic Regression [class balancing]','Logistic Regression [without class balancing]', 'Support Vector Machines', 'Random Forest [One Hot]', 'Random Forest [Response Coding]', 'Stacking [LR, SVM, NB] with LR', 'Stacking [LR,Linear,RF] with LR', 'Stacking [NB, LR, RF] with LR', 'Stacking [K-NN, NB, Linear] with LR', 'Stacking [LR, K-NN, Linear] with LR', 'Stacking [LR, Linear] with LR', 'Maximum Voting Classifier'],
    "Train (%)": [0.9,0.61,0.57,0.55,0.68,0.63,0.04,0.63,0.60,0.68,0.74,0.62,0.63,0.84],
    "CV (%)": [1.23,1.05,1.15,1.18,1.23,1.18,1.37,1.20,1.09,1.10,1.10,1.08,1.10,1.16],
    "Test (%)": [1.19,1.05,1.09,1.11,1.16,1.15,1.36,1.15,1.12,1.12,1.12,1.12,1.13,1.16],
    "Misclassification (%)": [43.23,35.90,37.59,36.65,39.28,41.54,57.70,37.74,36.39,35.93,36.09,36.39,36.39,38.79],
}

)
dataframe

Maching Learning Model  Train (%)  CV (%)  \
0                                     Naive Bayes       0.90    1.23   
1                             KNearest Neighbours       0.61    1.05   
2           Logistic Regression [class balancing]       0.57    1.15   
3   Logistic Regression [without class balancing]       0.55    1.18   
4                         Support Vector Machines       0.68    1.23   
5                         Random Forest [One Hot]       0.63    1.18   
6                 Random Forest [Response Coding]       0.04    1.37   
7                  Stacking [LR, SVM, NB] with LR       0.63    1.20   
8                 Stacking [LR,Linear,RF] with LR       0.60    1.09   
9                   Stacking [NB, LR, RF] with LR       0.68    1.10   
10            Stacking [K-NN, NB, Linear] with LR       0.74    1.10   
11            Stacking [LR, K-NN, Linear] with LR       0.62    1.08   
12                  Stacking [LR, Linear] with LR       0.63    1.10   
13                      Maximum Voting Classifier       0.84    1.16   

    Test (%)  Misclassification (%)  
0       1.19                  43.23  
1       1.05                  35.90  
2       1.09                  37.59  
3       1.11                  36.65  
4       1.16                  39.28  
5       1.15                  41.54  
6       1.36                  57.70  
7       1.15                  37.74  
8       1.12                  36.39  
9       1.12                  35.93  
10      1.12                  36.09  
11      1.12                  36.39  
12      1.13                  36.39  
13      1.16                  38.79

# 5.3 TF-IDF with Top 1K Features

In [3]:
import pandas as pd
dataframe = pd.DataFrame(
{
    "Maching Learning Model": ['Naive Bayes','K' 'Nearest Neighbours','Logistic Regression [class balancing]','Logistic Regression [without class balancing]', 'Support Vector Machines', 'Random Forest [One Hot]', 'Random Forest [Response Coding]', 'Stacking [LR, SVM, NB] with LR', 'Maximum Voting Classifier'],
    "Train (%)": [0.78,0.62,0.58,0.56,0.66,0.83,0.07,0.80,0.94],
    "CV (%)": [1.19,1.08,1.05,1.07,1.10,1.18,1.37,1.16,1.21],
    "Test (%)": [1.20,1.04,1.06,1.09,1.12,1.18,1.40,1.14,1.19],
    "Misclassification (%)": [38.15,37.59,37.03,37.21,36.65,40.60,50.75,37.14,38.04],
}

)
dataframe

Maching Learning Model  Train (%)  CV (%)  Test (%)  \
0                                    Naive Bayes       0.78    1.19      1.20   
1                            KNearest Neighbours       0.62    1.08      1.04   
2          Logistic Regression [class balancing]       0.58    1.05      1.06   
3  Logistic Regression [without class balancing]       0.56    1.07      1.09   
4                        Support Vector Machines       0.66    1.10      1.12   
5                        Random Forest [One Hot]       0.83    1.18      1.18   
6                Random Forest [Response Coding]       0.07    1.37      1.40   
7                 Stacking [LR, SVM, NB] with LR       0.80    1.16      1.14   
8                      Maximum Voting Classifier       0.94    1.21      1.19   

   Misclassification (%)  
0                  38.15  
1                  37.59  
2                  37.03  
3                  37.21  
4                  36.65  
5                  40.60  
6                  50.75  
7                  37.14  
8                  38.04

# 5.4 Appying Logistic Regression with Count Vectorizer (Unigrams & Bigrams)

In [6]:
import pandas as pd
dataframe = pd.DataFrame(
{
    "Maching Learning Model": ['Logistic Regression [class balancing]','Logistic Regression [without class balancing]'],
    "Train (%)": [0.87, 0.86],
    "CV (%)": [1.19, 1.21],
    "Test (%)": [1.258, 1.254],
    "Misclassification (%)": [37.21, 37.03],
}

)
dataframe

Maching Learning Model  Train (%)  CV (%)  Test (%)  \
0          Logistic Regression [class balancing]       0.87    1.19     1.258   
1  Logistic Regression [without class balancing]       0.86    1.21     1.254   

   Misclassification (%)  
0                  37.21  
1                  37.03

# 5.4.1 Average Weighted Word2VEC Tabular Result

In [8]:
import pandas as pd
dataframe = pd.DataFrame(
{
    "Maching Learning Model": ['Decision Trees','K' 'Nearest Neighbours','Logistic Regression [class balancing]','Logistic Regression [without class balancing]', 'Support Vector Classification', 'Support Vector Machines', 'Random Forest', 'Stacking [LR,SVC,NB] with K-NN', 'Stacking [Linear, K-NN, LR] with RF', 'Stacking [K-NN, SVC] with RF', 'Maximum Voting Classifier'],
    "Train (%)": [0.90,1.02,1.28,1.30,1.09,1.31,0.64,0.43,0.43,0.45,0.99],
    "CV (%)": [1.39,1.23,1.31,1.29,1.27,1.38,1.08,8.56,1.20,1.05,1.17],
    "Test (%)": [1.41,1.29,1.34,1.30,1.19,1.37,1.21,8.04,1.14,1.08,1.26],
    "Misclassification (%)": [43.04,42.85,48.68,45.11,46.05,0.5,37.59,38.34,41.50,37.74,45.41],
}

)
dataframe

Maching Learning Model  Train (%)  CV (%)  \
0                                  Decision Trees       0.90    1.39   
1                             KNearest Neighbours       1.02    1.23   
2           Logistic Regression [class balancing]       1.28    1.31   
3   Logistic Regression [without class balancing]       1.30    1.29   
4                   Support Vector Classification       1.09    1.27   
5                         Support Vector Machines       1.31    1.38   
6                                   Random Forest       0.64    1.08   
7                  Stacking [LR,SVC,NB] with K-NN       0.43    8.56   
8             Stacking [Linear, K-NN, LR] with RF       0.43    1.20   
9                    Stacking [K-NN, SVC] with RF       0.45    1.05   
10                      Maximum Voting Classifier       0.99    1.17   

    Test (%)  Misclassification (%)  
0       1.41                  43.04  
1       1.29                  42.85  
2       1.34                  48.68  
3       1.30                  45.11  
4       1.19                  46.05  
5       1.37                   0.50  
6       1.21                  37.59  
7       8.04                  38.34  
8       1.14                  41.50  
9       1.08                  37.74  
10      1.26                  45.41

# 5.4.2 TF-IDF Average Weighted Word2VEC Tabular Result

In [9]:
import pandas as pd
dataframe = pd.DataFrame(
{
    "Maching Learning Model": ['Decision Trees','K' 'Nearest Neighbours','Logistic Regression [class balancing]','Logistic Regression [without class balancing]', 'Support Vector Classification', 'Support Vector Machines', 'Random Forest', 'Stacking [LR,SVC,NB] with K-NN', 'Stacking [Linear, K-NN, LR] with RF', 'Stacking [K-NN, SVC] with RF', 'Maximum Voting Classifier'],
    "Train (%)": [1.83,1.83,1.83,1.83,1.83,1.83,1.83,8.67,1.83,1.83,1.83],
    "CV (%)": [1.83,1.83,1.83,1.83,1.83,1.83,1.83,8.66,1.83,1.83,1.83],
    "Test (%)": [1.83,1.83,1.83,1.83,1.83,1.83,1.83,8.69,1.83,1.83,1.83],
    "Misclassification (%)": [71.24,71.24,71.24,71.24,71.24,71.24,71.24,86.31,71.27,71.27,71.27],
}

)
dataframe

Maching Learning Model  Train (%)  CV (%)  \
0                                  Decision Trees       1.83    1.83   
1                             KNearest Neighbours       1.83    1.83   
2           Logistic Regression [class balancing]       1.83    1.83   
3   Logistic Regression [without class balancing]       1.83    1.83   
4                   Support Vector Classification       1.83    1.83   
5                         Support Vector Machines       1.83    1.83   
6                                   Random Forest       1.83    1.83   
7                  Stacking [LR,SVC,NB] with K-NN       8.67    8.66   
8             Stacking [Linear, K-NN, LR] with RF       1.83    1.83   
9                    Stacking [K-NN, SVC] with RF       1.83    1.83   
10                      Maximum Voting Classifier       1.83    1.83   

    Test (%)  Misclassification (%)  
0       1.83                  71.24  
1       1.83                  71.24  
2       1.83                  71.24  
3       1.83                  71.24  
4       1.83                  71.24  
5       1.83                  71.24  
6       1.83                  71.24  
7       8.69                  86.31  
8       1.83                  71.27  
9       1.83                  71.27  
10      1.83                  71.27

# 5.4.3 TF-IDF Top 2K Features with (n_gram = 1,4) Tabular Result

In [2]:
import pandas as pd
dataframe = pd.DataFrame(
{
    "Maching Learning Model": ['Naive Bayes','KNearest Neighbours','Logistic Regression [class balancing]','Logistic Regression [without class balancing]', 'Support Vector Machines', 'Random Forest [One Hot]', 'Random Forest [Response Coding]', 'Stacking [LR, SVM, NB] with LR', 'Stacking [LR,Linear,RF] with LR', 'Stacking [NB, LR, RF] with LR', 'Stacking [K-NN, NB, Linear] with LR', 'Stacking [LR, K-NN, Linear] with LR', 'Stacking [LR, Linear] with LR', 'Maximum Voting Classifier'],
    "Train (%)": [0.58,0.63,0.45,0.45,0.59,0.87,0.05,0.38,0.46,0.43,0.44,0.42,0.43,0.88],
    "CV (%)": [1.232,1.060,1.026,1.056,1.072,1.191,1.293,1.083,1.057,1.089,1.090,1.060,1.072,1.067],
    "Test (%)": [1.177,1.041,0.943,0.962,0.996,1.135,1.246,1.15,1.112,1.158,1.169,1.131,1.135,1.039],
    "Misclassification (%)": [40.03,37.78,34.39,34.96,35.52,43.23,50.37,36.69,35.48,38.64,39.39,35.63,34.73,31.87],
}

)
dataframe

Maching Learning Model  Train (%)  CV (%)  \
0                                     Naive Bayes       0.58   1.232   
1                             KNearest Neighbours       0.63   1.060   
2           Logistic Regression [class balancing]       0.45   1.026   
3   Logistic Regression [without class balancing]       0.45   1.056   
4                         Support Vector Machines       0.59   1.072   
5                         Random Forest [One Hot]       0.87   1.191   
6                 Random Forest [Response Coding]       0.05   1.293   
7                  Stacking [LR, SVM, NB] with LR       0.38   1.083   
8                 Stacking [LR,Linear,RF] with LR       0.46   1.057   
9                   Stacking [NB, LR, RF] with LR       0.43   1.089   
10            Stacking [K-NN, NB, Linear] with LR       0.44   1.090   
11            Stacking [LR, K-NN, Linear] with LR       0.42   1.060   
12                  Stacking [LR, Linear] with LR       0.43   1.072   
13                      Maximum Voting Classifier       0.88   1.067   

    Test (%)  Misclassification (%)  
0      1.177                  40.03  
1      1.041                  37.78  
2      0.943                  34.39  
3      0.962                  34.96  
4      0.996                  35.52  
5      1.135                  43.23  
6      1.246                  50.37  
7      1.150                  36.69  
8      1.112                  35.48  
9      1.158                  38.64  
10     1.169                  39.39  
11     1.131                  35.63  
12     1.135                  34.73  
13     1.039                  31.87

# 6. Observations

*As we can observe from the above tabular results, I've applied various techniques to reduce the test and cv erro below 1*

*I've applied various feature generation techniques like Bag of words, TF-IDF, Average Word2Vec, TF-IDF Weighted Word2Vec on cancer diagnosis dataset and printed the results in a tabular form*


1. **Bag of Words**:-

*Stacking Models have obtained better results when compare to other classifiers. Among them the best result we got by stacking Logistic Regression, Linear Regression and Random Forest and using Logistic Regression as the classifier*.


Train Error: 0.65,
CV Error   : 1.14,
Test Error : 1.08,
Misclassification Error : 35.33


*Apart from Stacking models, K-NN also obtained a good result*.

Train Error: 0.70,
CV Error   : 1.10,
Test Error : 1.09,
Misclassification Error : 39.66


2. **Term Frequency - Inverse Document Frequency (TF-IDF)**:-

*K Nearest Neighbours have obtained the best result*


Train Error: 0.61,
CV Error   : 1.05,
Test Error : 1.05,
Misclassification Error : 35.90


*The second best result is from Logistic Regression [with class balancing]*

Train Error: 0.57,
CV Error   : 1.15,
Test Error : 1.09,
Misclassification Error : 37.59


3. **TF-IDF with Top 1k Features**

*K Nearest Neighbours have obtained the best result*

Train Error: 0.62,
CV Error   : 1.08,
Test Error : 1.04,
Misclassification Error : 37.59


*The second best result is from Logistic Regression [with class balancing]*

Train Error: 0.58,
CV Error   : 1.05,
Test Error : 1.06,
Misclassification Error : 37.03


4. **Logistic Regression with Count Vectorizer (Unigrams & Bigrams)**

*We've applied logistic regression with count vectorizer including both unigrams and bigrams*.

*Logistic Regression without class balancing result is some what better when compare to logistic regression with class balancing result*

Train Error: 0.86,
CV Error   : 1.21,
Test Error : 1.25,
Misclassification Error : 37.03

5. **Average Weighted Word2VEC**


*Stacking K Nearest Neighbours & Support Vector Machines and Random Forest as Classifier, we obtained the best result compare to other classifiers*.

Train Error: 0.45,
CV Error   : 1.05,
Test Error : 1.08,
Misclassification Error : 37.74


6. **TF-IDF Top 2K Features with (n_gram = 1,4)**


*Among all the classifiers/models we've used, TF-IDF with top 2k features including ngrams (1,4) got the best result reducing the test error to below 1. Below are the top three best results*


> Logistic Regression with Class Balancing

Train Error: 0.45,
CV Error   : 1.02,
Test Error : 0.94,
Misclassification Error : 34.39


> Logistic Regression without class balancing


Train Error: 0.45,
CV Error   : 1.05,
Test Error : 0.96,
Misclassification Error : 34.96


> Support Vector Machines


Train Error: 0.59,
CV Error   : 1.07,
Test Error : 0.99,
Misclassification Error : 35.52